internal/cache/async.go

```diff
@@ -0,0 +1,155 @@
+	queue              store.ShardedUniqueQueue
+	objectStore        store.ObjectStore
+}
+
+func (as *asyncClient) Run(ctx context.Context) {
+	for _, q := range as.queue.GetConsumers() {
+		go as.runWorker(ctx, q)
+	}
+}
+
+func (as *asyncClient) runWorker(ctx context.Context, requests <-chan func() store.Request) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case requestGetter := <-requests:
+			r := requestGetter()
+			switch r.Type {
+			case store.CreateRequestType:
+				as.doCreate(requestCtx(ctx, r.Key, "create"), r.Key)
+			case store.UpdateRequestType:
+				as.doUpdate(requestCtx(ctx, r.Key, "update"), r.Key)
+			case store.DeleteRequestType:
+				as.doDelete(requestCtx(ctx, r.Key, "delete"), r.Key)
+			}
+		}
+	}
+}
+
+func (as *asyncClient) doCreate(ctx context.Context, key store.Key) {
+	obj, ok := as.objectStore.Get(key)
+	if !ok {
+		svc1log.FromContext(ctx).Info("Ignoring request for deleted object")
+		return
+	}
+	result := as.emptyObjectCreator()
+	err := as.client.Post().
+		Namespace(obj.GetNamespace()).
+		Resource(as.resourceName).
+		Body(obj).
+		Do().
+		Into(result)
+	switch {
+	case err == nil:
+		as.objectStore.OverrideResourceVersionIfNewer(ctx, obj)
+	case isRetryableError(ctx, err):
+		as.queue.AddIfAbsent(store.CreateRequest(obj))
+	default:
+		logAndIgnore(ctx, err)
+	}
+}
+
+func (as *asyncClient) doUpdate(ctx context.Context, key store.Key) {
+	obj, ok := as.objectStore.Get(key)
+	if !ok {
+		svc1log.FromContext(ctx).Info("Ignoring request for deleted object")
+		return
+	}
+
+	result := as.emptyObjectCreator()
+	err := as.client.Put().
+		Namespace(obj.GetNamespace()).
+		Resource(as.resourceName).
+		Name(obj.GetName()).
+		Body(obj).
+		Do().
+		Into(result)
+	if err != nil {
+		as.objectStore.OverrideResourceVersionIfNewer(ctx, obj)
+	}
+	switch {
+	case err == nil:
+		as.objectStore.OverrideResourceVersionIfNewer(ctx, obj)
+	case isRetryableError(ctx, err):
+		as.queue.AddIfAbsent(store.UpdateRequest(obj))
+	case errors.IsConflict(err):
+		svc1log.FromContext(ctx).Warn("got conflict, will retry", svc1log.Stacktrace(err))
+		as.queue.AddIfAbsent(store.UpdateRequest(obj))
+	default:
+		logAndIgnore(ctx, err)
+	}
+}
+
+func (as *asyncClient) doDelete(ctx context.Context, key store.Key) {
+	err := as.client.Delete().
+		Namespace(key.Namespace).
+		Resource(as.resourceName).
+		Name(key.Name).
+		Do().
+		Error()
+	if err != nil {
+		as.objectStore.Delete(key)
+	}
+	switch {
+	case err == nil:
+		as.objectStore.Delete(key)
+	case isRetryableError(ctx, err):
+		as.queue.AddIfAbsent(store.Request{Key: key, Type: store.DeleteRequestType})
+	default:
+		logAndIgnore(ctx, err)
+	}
+}
+
+func isRetryableError(ctx context.Context, err error) bool {
+	isRetryable := errors.IsServerTimeout(err) || errors.IsServiceUnavailable(err) ||
+		errors.IsTooManyRequests(err) || errors.IsTimeout(err)
+	if isRetryable {
+		svc1log.FromContext(ctx).Warn("got retryable error, will retry", svc1log.Stacktrace(err))
```

Comment:
nit - I'm not a fan of logging `will do X` in a function that just returns a boolean, who knows if the caller will do the right thing.

Instead, maybe factor out `as.queue.AddIfAbsent(...)` as a function called `addTaskToRetry(...)` and move that log statement there?